src/experiments.js

```diff
@@ -47,6 +51,102 @@ export function isCanary(win) {
 }
 
 
+function getExperimentsKey_() {
+  let key;
+  const url = '' //TODO(kmh287): fill this in
+  if (url) {
+    return xhrFor(this.win).fetchJson(url, {
+      mode: 'cors',
+      method: 'GET',
+      ampCors: false,
+      credentials: 'omit',
+    }).then(jwkObj => {
+      if (isObject(jwkObj)) {
+        return this.crypto_.importPublicKey('experiments', key);
+      } else {
+        throw new Error(`Invalid response from experiments signing server`);
+      }
+    });
+  } else {
+    // The experiment signing service url wasn't found
+    return Promise.reject(
+        new Error(`Signing service for experiments does not exist.`));
+  }
+}
+
+/**
+ * Whether the specified experiment is on or off for origin trials. You should
+ * check if the experiment is already enabled before calling this function.
+ * @param {!Window} win
+ * @param {string} experimentId
+ * @return {!Promise<boolean>}
+ */
+export function isExperimentOnForOriginTrial(win, experimentId) {
+  const crypto = cryptoFor(win);
+  const meta =
+      win.document.head.querySelector('meta[name="amp-experiment-token"]');
+  if (!meta || !crypto.isCryptoAvailable()) {
+    return Promise.resolve(false);
+  }
+  const token = meta.getAttribute('content');
+  if (!token) {
+    return Promise.reject(new Error('Unable to read experiments token'));
+  }
+  /**
+   * token = encode64(version + length + config + sign(config, private_key))
+   * version = 1 byte version of the token format (starting at 0x0)
+   */
+  const bytes = stringToBytes(atob(token));
+  const version = bytes[0];
+  if (version === 0) {
+   /**
+    * Version 0:
+    * length = 4 bytes representing number of bytes in config
+    * config = string containing the experiment ID, origin URL, etc.
+    */
+    const configLen = bytesToInt(bytes.subarray(1,5));
+    if (configLen > bytes.length - 5) {
+      return Promise.reject(
+          new Error('Specified len extends past end of buffer'));
+    }
+    const configBytes = bytes.subarray(5, 5 + configLen);
+    const signatureBytes = bytes.subarray(5 + configLen);
+
+    return getExperimentsKey_().then(keyInfo => {
+      return crypto.verifySignature(configBytes, signatureBytes, keyInfo)
+    }).then(verified => {
+      if (!verified) {
+        throw new Error('Failed to verify config signature');
+      }
+      const configStr = utf8DecodeSync(configBytes);
+      const config = JSON.parse(configStr);
+
+      const approvedOrigin = config['origin'];
+      const url = ampdocServiceFor(win).getAmpDoc().getUrl();
+      const sourceOrigin = getSourceOrigin(url);
+      if (approvedOrigin !== sourceUrl) {
+        throw new Error('Config does not match current origin');;
+      }
+      const experiments = config['experiments'];
+      const experiment = experiments[experimentId];
```

Comment:
When we generate these tokens, we're going to need to either keep track of all experiments for each origin on our end, or request that all experiments be requested at once. Alternatively we can allow multiple `meta` tags with tokens, but that sounds like it could get slow very quickly.